nixos/pulseaudio: remove broken support for 32bit on 64bit systems

ivanbrennan opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

In #154276, alsa-lib was updated and old patches removed. One of those patches, from 2014, allowed alsa-lib to recognize a `libs` field in /etc/asound.conf, and the pulseaudio module leveraged this to optionally add support for running both 32bit and 64bit apps on 64bit platforms.

With the patch removed, alsa-lib is unable to parse the /etc/asound.conf generated by the pulseaudio module (regardless whether the user has opted into 32bit support), and alsa utils such as `alsactl`, `alsamixer`, `speaker-test`, are broken. E.g.
```
$ alsactl monitor default
alsa-lib control.c:1464:(snd_ctl_open_conf) Unknown field libs
Cannot open ctl default
```

Once the issue was identified, it was reasoned that the 32bit support is too niche a use-case to justify patching a system as complex as alsa, especially given the availability of pipewire.

I removed the `support32Bit` option from pulseaudio and changed the generated config to use syntax recognized by alsa-lib. I then removed several references to this option from other modules (steam, jack, pipewire).

***Note***: I rebuilt my system against this branch and confirmed that alsa works as expected for my use-cases. I have not tested steam, jack, or pipewire.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

ivanbrennan committed:
nixos/pulseaudio: remove unparsable 32bit config

Back in 2014, the following commits patched alsa-lib to support running
apps with 32bit sound on 64bit platforms, and configured
/etc/asound.conf to leverage that support:
ab8ef63ff49dd6276f765cce10d88b3e5b86b837
0c8ad65560fa1df07ea7b7937e02b44b1ce2b498

A recent commit updated alsa-lib (1.2.5.1 -> 1.2.6.1), and removed the
patch, but left the pulseaudio module unchanged, meaning it continued to
generate an etc/asound.conf containing the `libs` field, which alsa-lib
cannot recognize (the patch was providing support for that field).
aeea1bb53b28fc7bbe4583bd21f5bda8b05d5041

As a result, a system whose configuration enables pulseaudio can no
longer run alsa utils, regardless of whether support32Bit is enabled.
E.g.

  $ alsactl monitor default
  alsa-lib control.c:1464:(snd_ctl_open_conf) Unknown field libs
  Cannot open ctl default

  $ speaker-test -t wav -c 2

  speaker-test 1.2.6

  Playback device is default
  Stream parameters are 48000Hz, S16_LE, 2 channels
  WAV file(s)
  ALSA lib pcm.c:2576:(snd_pcm_open_conf) Unknown field libs
  Playback open error: -22,Invalid argument

Once the issue was identified, it was reasoned that the 32bit support is
too niche a use-case to justify patching a system as complex as alsa,
especially in 2022 and with pipewire providing a viable alternative
solution.

Remove the unused (and now unusable) 32bit configuration code from
/etc/asound.conf, so alsa-lib can again parse it and run properly.

ivanbrennan committed:
remove references to pulseaudio.support32Bit

As described in the previous commit, this option was removed. Remove
references to it, and produce etc/asound.conf files that alsa-lib is
able to parse.

ivanbrennan commented:
@L-as You mentioned,
> The only scenario I can think of where this is still relevant is Steam with proprietary old 32-bit games that use ALSA directly, but you can just switch to PipeWire for that! Perhaps that's just something we should note in the release notes for the next NixOS release.

I'm not sure what action the notes should suggest to affected users. Could you advise?

Riey commented:
I also have this issue in current master (not pulseaudio but pipewire/alsa.enable = true)

with this patch, I check speaker-test alsamixer and some steam games(Hollow Knight and Oxygen Not included) and works fine

Synthetica9 commented:
... Now I'm confused about what `hardware.pulseaudio.support32Bit` did? I just rebased #153916 on master and commented out `support32Bit` but the 32 bit portion of the test still works...

L-as commented:
Can someone try this with a 32-bit game?

jansol reviewed (changes requested):
The alsa 32bit options have been pretty widely recommended in the past due to steam so I'd like to see a proper deprecation error instead of just silently removing the options.

The pipewire portion of the diff looks good to me otherwise.

ivanbrennan committed:
nixos/pulseaudio: support32Bit no longer supported

Use an assertion to display a proper error message if the user's config
has the `support32Bit` option enabled. Example error output:

  error:
         Failed assertions:
         - `hardware.pulseaudio.support32Bit` is no longer supported.

ivanbrennan commented on nixos/modules/config/pulseaudio.nix:
```diff
@@ -210,6 +218,11 @@ in {
     (mkIf cfg.enable {
+      assertions = [{
+        assertion = !cfg.support32Bit;
+        message = "`hardware.pulseaudio.support32Bit` is no longer supported.";
```
> The alsa 32bit options have been pretty widely recommended in the past due to steam so I'd like to see a proper deprecation error instead of just silently removing the options.

@jansol That makes sense. This commit adds a clearer message, but did you mean an actual deprecation process, where we fix the currently broken 32bit support and provide a deprecation warning about its eventual removal? That would require reintroducing the patch that was removed in #154276, but I'm not opposed to doing that if it's warranted (and it sounds like it is).

I'd also like to provide a more informative message than I've done in this commit, something like,
```
In order to support 32bit audio on a 64bit system, please use ??? instead.
```
But I don't know enough about this domain to do so. Is there a pipewire-related suggestion that would be appropriate in place of `???` above?
cc @L-as

ivanbrennan committed:
assertions: support32Bit no longer supported

starcraft66 commented:
Tested this branch on my system and it fixed pipewire support for ALSA within Mixxx.

dtzWill commented:
Thanks for working on this!

Small note: this also drops (similarly broken) 32bit alsa support for jack and pipewire (removing the `libs` entries from those asoundrc's as well), perhaps update the PR title?

ivanbrennan commented:
> Small note: this also drops (similarly broken) 32bit alsa support for jack and pipewire (removing the `libs` entries from those asoundrc's as well), perhaps update the PR title?

@dtzWill That's a good point, but I'm considering changing the direction of this PR to fix the 32bit support (re-adding the alsa-lib patch that was removed) and deprecate the option rather than removing it right away. (see https://github.com/NixOS/nixpkgs/pull/157352#discussion_r795285011)

I'll update the PR title one way or the other though :+1:

ivanbrennan commented:
Given that it's not clear what alternative to suggest to users who currently rely on the 32bit support, I think the best thing is to fix what's currently broken, and leave the deprecation/removal of 32bit support to someone knowledgeable enough to compose appropriate deprecation warnings (including suggested config changes for affected users).

So I've created an alternative PR that confines itself to fixing what's currently broken: #157631

ivanbrennan commented:
Closing this in favor of #157631
